iceoryx2-services/discovery/src/service_discovery/service.rs

```diff
@@ -0,0 +1,331 @@
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            sync_on_initialization: true,
+            include_internal: true,
+            publish_events: true,
+            max_subscribers: 10,
+            send_notifications: true,
+            max_listeners: 10,
+        }
+    }
+}
+
+/// The service discovery service.
+///
+/// This service is responsible for tracking and publishing information about
+/// services in the system. It can detect when services are added or removed,
+/// and notify interested parties about these changes.
+///
+/// # Type Parameters
+///
+/// * `S` - The service type that this discovery service operates on.
+pub struct Service<S: ServiceType> {
+    discovery_config: Config,
+    iceoryx_config: IceoryxConfig,
+    _node: Node<S>,
+    publisher: Option<Publisher<S, [u8], ()>>,
+    notifier: Option<Notifier<S>>,
+    tracker: Tracker<S>,
+}
+
+impl<S: ServiceType> Service<S> {
+    /// Creates the service discovery service.
+    ///
+    /// This function initializes a new service discovery service with the provided
+    /// configuration. The service will track services of type `S` in the system.
+    ///
+    /// # Parameters
+    ///
+    /// * `discovery_config` - Configuration for the discovery service.
+    /// * `iceoryx_config` - Configuration for the underlying iceoryx system.
+    ///
+    /// # Returns
+    ///
+    /// A result containing either the created service or an error if creation failed.
+    ///
+    /// # Errors
+    ///
+    pub fn create(
+        discovery_config: &Config,
+        iceoryx_config: &IceoryxConfig,
+    ) -> Result<Self, CreationError> {
+        let node = NodeBuilder::new()
+            .config(iceoryx_config)
+            .create::<S>()
+            .map_err(|_| CreationError::NodeCreationFailure)?;
+
+        let mut publisher = None;
+        if discovery_config.publish_events {
+            let publish_subscribe = node
+                .service_builder(service_name())
+                .publish_subscribe::<Payload>()
+                .subscriber_max_borrowed_samples(10)
+                .history_size(10)
+                .subscriber_max_buffer_size(10)
+                .max_publishers(1)
+                .max_subscribers(discovery_config.max_subscribers)
+                .create()
+                .map_err(|e| {
+                    match e {
+                        iceoryx2::service::builder::publish_subscribe::PublishSubscribeCreateError::AlreadyExists => CreationError::ServiceAlreadyExists,
+                        iceoryx2::service::builder::publish_subscribe::PublishSubscribeCreateError::IsBeingCreatedByAnotherInstance => CreationError::ServiceAlreadyExists,
+                        _ => CreationError::ServiceCreationFailure,
+                    }
+                })?;
+
+            publisher = Some(
+                publish_subscribe
+                    .publisher_builder()
+                    .initial_max_slice_len(1024)
```

Comment:
This should hopefully be removed when `StaticConfig` is `ZeroCopySend`.

Reply:
Added to: https://github.com/eclipse-iceoryx/iceoryx2/issues/708